core/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -928,22 +907,24 @@ private void assertCanWrite() throws IOException {
         for (Path path : nodeDataPaths()) { // check node-paths are writable
             tryWriteTempFile(path);
         }
-        for (String index : this.findAllIndices()) {
-            for (Path path : this.indexPaths(index)) { // check index paths are writable
-                Path statePath = path.resolve(MetaDataStateFormat.STATE_DIR_NAME);
-                tryWriteTempFile(statePath);
-                tryWriteTempFile(path);
-            }
-            for (ShardId shardID : this.findAllShardIds(new Index(index, IndexMetaData.INDEX_UUID_NA_VALUE))) {
-                Path[] paths = this.availableShardPaths(shardID);
-                for (Path path : paths) { // check shard paths are writable
-                    Path indexDir = path.resolve(ShardPath.INDEX_FOLDER_NAME);
-                    Path statePath = path.resolve(MetaDataStateFormat.STATE_DIR_NAME);
-                    Path translogDir = path.resolve(ShardPath.TRANSLOG_FOLDER_NAME);
-                    tryWriteTempFile(indexDir);
-                    tryWriteTempFile(translogDir);
-                    tryWriteTempFile(statePath);
-                    tryWriteTempFile(path);
+        for (String indexFolderName : this.availableIndexFolders()) {
```

Comment:
Now `availableIndexFolders` doesn't filter out folders with no state file in them rather on upgrade we throw a hard exception whenever we find a folder without a valid state.